build-logic/src/main/kotlin/dokkabuild/utils/SystemPropertyAdder.kt

```diff
@@ -0,0 +1,146 @@
+    get() {
+        val spa = extensions.findByType<SystemPropertyAdder>()
+            ?: extensions.create<SystemPropertyAdder>("SystemPropertyAdder", this)
+        return spa
+    }
+
+
+abstract class SystemPropertyAdder @Inject internal constructor(
+    private val task: Test,
+) {
+    fun inputDirectory(
+        key: String,
+        value: Directory,
+    ): TaskInputFilePropertyBuilder {
+        task.jvmArgumentProviders.add(
+            SystemPropertyArgumentProvider(key, value) {
+                it.asFile.invariantSeparatorsPath
+            }
+        )
+        return task.inputs.dir(value)
+            .withPropertyName("SystemProperty input directory $key")
+    }
+
+    fun inputFile(
+        key: String,
+        file: RegularFile,
+    ): TaskInputFilePropertyBuilder {
+        task.jvmArgumentProviders.add(
+            SystemPropertyArgumentProvider(key, file) {
+                it.asFile.invariantSeparatorsPath
+            }
+        )
+        return task.inputs.file(file)
+            .withPropertyName("SystemProperty input file $key")
+    }
+
+    fun inputFile(
+        key: String,
+        file: Provider<out RegularFile>,
+    ): TaskInputFilePropertyBuilder {
+        task.jvmArgumentProviders.add(
+            SystemPropertyArgumentProvider(key, file) {
+                it.orNull?.asFile?.invariantSeparatorsPath
+            }
+        )
+        return task.inputs.file(file)
+            .withPropertyName("SystemProperty input file $key")
+    }
+
+    fun inputFiles(
+        key: String,
+        files: Provider<out FileCollection>,
+    ): TaskInputFilePropertyBuilder {
+        task.jvmArgumentProviders.add(
+            SystemPropertyArgumentProvider(key, files) { it.orNull?.asPath }
+        )
+        return task.inputs.files(files)
+            .withPropertyName("SystemProperty input files $key")
+    }
+
+    fun inputProperty(
+        key: String,
+        value: Provider<out String>,
+    ): TaskInputPropertyBuilder {
+        task.jvmArgumentProviders.add(
+            SystemPropertyArgumentProvider(key, value) { it.orNull }
+        )
+        return task.inputs.property("SystemProperty input property $key", value)
+    }
+
+    @JvmName("inputBooleanProperty")
+    fun inputProperty(
+        key: String,
+        value: Provider<out Boolean>,
+    ): TaskInputPropertyBuilder = inputProperty(key, value.map { it.toString() })
+
+    /**
+     * Add a System Property (in the format `-D$key=$value`).
+     *
+     * [value] will be treated as if it were annotated with [org.gradle.api.tasks.Internal]
+     * and will _not_ be registered as a Gradle [org.gradle.api.Task] input.
+     * (Which is beneficial in cases where [value] is optional, or is not reproducible,
+     * because such a property might disrupt task caching.)
+     *
+     * If you want to register the property as a Task input, use the
+     * [Test.systemProperty][dokkabuild.utils.systemProperty] above instead.
```

Comment:
should be rephrased now

Reply:
Sorry, how do you think it could be rephrased?

Reply:
I mean, it's now a function inside an object which is created by `[Test.systemProperty]`, so may be suggesting just using other property will be more appropriate here, like:
```suggestion
     * [inputProperty] above instead.
```

Reply:
Good point, thanks for explaining. I've updated the kdoc.